### ISSUE TITLE
Downcasting from fp to uint formats diverges from numpy behavior

### ISSUE THREAD

**Issue by erdmann** [google/jax#10005]
`jax.numpy` differs from `numpy` in the way it treats downcasting from floating-point formats to unsigned integer formats (and possibly with other combinations).

Consider the following:

```python
import numpy as np
import jax.numpy as jnp

i32 = np.array([-255, -1, 256, 255, 257], dtype=np.int32)
f32 = np.array([-255, -1, 256, 255, 257], dtype=np.float32)

print(f'                 {i32.astype(np.uint8) = }')
print(f'{jnp.array(i32).astype(np.uint8) = }')
print()
print(f'                 {f32.astype(np.uint8) = }')
print(f'{jnp.array(f32).astype(np.uint8) = }')

#                  i32.astype(np.uint8) = array([  1, 255,   0, 255,   1], dtype=uint8)
# jnp.array(i32).astype(np.uint8) = DeviceArray([  1, 255,   0, 255,   1], dtype=uint8)
#
#                  f32.astype(np.uint8) = array([  1, 255,   0, 255,   1], dtype=uint8)
# jnp.array(f32).astype(np.uint8) = DeviceArray([  0,   0, 255, 255, 255], dtype=uint8)
```
My expectation was that the last line would match the others, but JAX is clipping rather than wrapping around like the `numpy` 
behavior in the penultimate line.  I looked for this in [JAX - The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html), but didn't turn up anything.   If this behavior is not a bug, it would be helpful to document it there.

Thanks in advance!

**Comment by YouJiacheng** [comment]
You may refer to:
https://github.com/google/jax/blob/e0d3946e341903bb5cedfe605618b06438f9fe1d/jax/_src/numpy/lax_numpy.py#L4882
=> `DeviceArray.astype` is implemented in `_astype`

https://github.com/google/jax/blob/e0d3946e341903bb5cedfe605618b06438f9fe1d/jax/_src/numpy/lax_numpy.py#L4343
=> `_astype` use `lax.convert_element_type`

https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.convert_element_type.html
=> `lax.convert_element_type` wrap XLA ConvertElementType

https://www.tensorflow.org/xla/operation_semantics#convertelementtype
> Similar to an element-wise static_cast in C++

> The precise float-to-int and visa-versa conversions are currently unspecified, but may become additional arguments to the convert operation in the future. Not all possible conversions have been implemented for all targets.

It seems that the behavior is implementation defined.

**Comment by erdmann** [comment]
Thanks, @YouJiacheng.   For future reference, this is on GPU:

```python
print(f'{jax.devices()[0].device_kind = }')
print(f'{jax.__version__ = }, {jaxlib.__version__ = }')

# jax.devices()[0].device_kind = 'Tesla V100-DGXS-32GB'
# jax.__version__ = '0.3.4', jaxlib.__version__ = '0.3.2'
```

**Comment by jakevdp** [comment]
Thanks for bringing this up. I've clarified the behavior in the docstring for `astype` in https://github.com/google/jax/pull/10009

**Comment by erdmann** [comment]
More often than not I do conversions using this form: `jnp.uint8(jnp.arange(300, dtype='<f4'))`.  Is there a good way to document the unexpected numpy divergence in that case?  Or is there some sort of warning behavior that can be set to protect against this footgun?

**Comment by erdmann** [comment]
For me the sequence in getting to be comfortable with JAX went something like this:

1.  Read all over the place that it is essentially a wicked fast drop-in replacement for numpy on the GPU.
2. Experiment with it and quickly discover that I couldn't modify arrays in-place.
3. Go read the documentation and, in particular, try to absorb all the teachings in [JAX - The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html).

So, when my JAX code started behaving differently than numpy in the sense that downcasting now cared about whether the source array was float or integer type, I just assumed that I had missed some sharp bits and I went back to study that carefully again.  I filed this bug when I didn't find anything about implementation-dependent casting behavior, because I think of the Sharp Bits document as being the one-stop shop for the ways in which JAX isn't just a drop-in replacement for numpy.

I agree that warnings would probably be too noisy, and it would probably be a lot of work to implement a JAX-wide "be pedantic" setting for warnings, although it would have saved me in this case.  An entry in [JAX - The Sharp Bits](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html) would've probably also saved me since I would've mentally logged that casting was a different-from-numpy danger zone.

Thanks for the very helpful response, by the way.  I think the support that the team gives to JAX users is outstanding.
